### PR TITLE
1. Tweaked the closest bar algo such that being within a bar's secondary...

### PR DESCRIPTION
**Pull request by istarnx**
... axis span sets its distance to 0.

2. Now test bbox intersection with a tolerance of zero: gives us stable behaviour when moving along the dominant axis at the same secondary axis value.
3. Tests. :)

### CODE REVIEW

File: test/components/plots/stackedBarPlotTests.ts [hunk 1]
```diff
@@ -89,6 +89,50 @@ describe("Plots", () => {
       assert.deepEqual(dataset2.data(), originalData2, "underlying data is not modified");
       svg.remove();
     });
+
+    it("returns correct closest plot data", () => {
```
Comment (bluong) [inline test/components/plots/stackedBarPlotTests.ts]: test name is not precise enough

File: src/components/plots/barPlot.ts [hunk 2]
```diff
@@ -179,12 +179,21 @@ export module Plot {
           var secondaryDist = 0;
 
           // if we're inside a bar, distance in both directions should stay 0
-          if (!_Util.Methods.intersectsBBox(queryPoint.x, queryPoint.y, barBBox)) {
+          if (!_Util.Methods.intersectsBBox(queryPoint.x, queryPoint.y, barBBox, 0)) {
```
Comment (bluong) [inline src/components/plots/barPlot.ts]: Why this change?

File: src/components/plots/barPlot.ts [hunk 3]
```diff
@@ -179,12 +179,21 @@ export module Plot {
...
-            var plotPtSecondary = this._isVertical ? plotPt.y : plotPt.x;
-
             primaryDist = Math.abs(queryPtPrimary - plotPtPrimary);
-            secondaryDist = Math.abs(queryPtSecondary - plotPtSecondary);
+
+            // compute this bar's min and max along the secondary axis
+            var barMinSecondary = this._isVertical ? barBBox.y : barBBox.x;
+            var barMaxSecondary = barMinSecondary + (this._isVertical ? barBBox.height : barBBox.width);
+
+            if (queryPtSecondary > barMinSecondary && queryPtSecondary <= barMaxSecondary) {
```
Comment (bluong) [inline src/components/plots/barPlot.ts]: Why > instead of >=

Comment (istarnx) [inline src/components/plots/barPlot.ts]: So, my intent was to tie-break using this approach, but after thinking about it again, I'm not sure. _intersectsBBox() will end up picking two bars if you're exactly on their border value (note the >= and <= in its comparison code) so perhaps it's reasonable to do the same thing here and let the user deal with multiple entities exactly the same distance apart.

 so it's reasonable that this logic do the same and leave it up to the user to deal with two

File: src/components/plots/barPlot.ts (hunk 2 above)
Comment (istarnx) [inline src/components/plots/barPlot.ts]: Note the commit comment: "Now test bbox intersection with a tolerance of zero: gives us stable behaviour when moving along the dominant axis at the same secondary axis value." 

It's to make sliding the mouse outside of a bar, but on the same secondary axis value, keep that bar as closest. I.e., when the logic switches from the "within bar" to "within bar's y-range" the resultant bar was different for the same y value when I slid the mouse along x.

**Comment by bluong** [comment]
Needs to obey tolerance

File: src/components/plots/barPlot.ts [hunk 4]
```diff
@@ -179,12 +184,21 @@ export module Plot {
...
-            var plotPtSecondary = this._isVertical ? plotPt.y : plotPt.x;
-
             primaryDist = Math.abs(queryPtPrimary - plotPtPrimary);
-            secondaryDist = Math.abs(queryPtSecondary - plotPtSecondary);
+
+            // compute this bar's min and max along the secondary axis
+            var barMinSecondary = this._isVertical ? barBBox.y : barBBox.x;
+            var barMaxSecondary = barMinSecondary + (this._isVertical ? barBBox.height : barBBox.width);
+
+            if (queryPtSecondary >= barMinSecondary - tolerance && queryPtSecondary <= barMaxSecondary + tolerance) {
```
Comment (bluong) [inline src/components/plots/barPlot.ts]: This makes sense from a code standpoint, but from a semantical standpoint, can we change it to have the same semantics as intersectsBBox?

aka

queryPtSecondary + tolerance >= barMinSecondary && queryPtSecondary - tolerance <= barMaxSecondary

This corresponds with the fact that the queryPt has an invisible rectangle around itself rather than the bbox of the bar having that invisible rectangle
